spark3/src/main/java/org/apache/spark/sql/connector/catalog/SupportsMetadataOnlyDeletes.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.connector.catalog;
+
+import org.apache.spark.sql.sources.Filter;
+
+// TODO: any better ideas? make it generic as SupportsDeleteUsingFilters and merge it into SupportsDelete upstream?
```

Comment:
@rdblue, I feel like `SupportsDelete` is very rudimentary and did not assume we can support deletes by rewriting plans. We will have to think how to approach that.

Reply:
My intent there was to first run `DeleteFrom` and if that failed, then attempt the non-metadata version.

We could try doing that by extending `OverwriteFiles` for the delete and calling `SupportsDelete.deleteWhere`. If that fails, then call `OverwriteFiles.doExecute`.